runtime/cudaq/platform/default/rest/helpers/iqm/IQMServerHelper.cpp

```diff
@@ -277,34 +299,208 @@ IQMServerHelper::generateRequestHeader() const {
       {"Connection", "keep-alive"},
       {"User-Agent", "cudaq/IQMServerHelper"},
       {"Accept", "*/*"}};
-  auto apiToken = readApiToken();
-  if (apiToken.has_value()) {
-    headers["Authorization"] = "Bearer " + apiToken.value();
-  };
+
+  // Prefer the authorization token set in the environment variable.
+  if (authToken.has_value()) {
+    headers["Authorization"] = "Bearer " + authToken.value();
+  } else {
+    // Fallback to authorization token from legacy JSON file.
+    auto apiToken = readApiToken();
+    if (apiToken.has_value()) {
+      headers["Authorization"] = "Bearer " + apiToken.value();
+    };
+  }
+
   return headers;
 }
 
 void IQMServerHelper::updatePassPipeline(
     const std::filesystem::path &platformPath, std::string &passPipeline) {
-  // Note: the leading and trailing single quotes are needed in case there are
-  // spaces in the filename.
   std::string pathToFile;
+
+  // Allow setting of a mapping file from outside.
   auto iter = backendConfig.find("mapping_file");
   if (iter != backendConfig.end()) {
-    // Use provided path to file
-    pathToFile = std::string("'") + iter->second + std::string("'");
+    // Use provided string as path+filename
+    pathToFile = iter->second;
   } else {
-    // Construct path to file
-    pathToFile =
-        std::string("'") +
-        std::string(platformPath / std::string("mapping/iqm") /
-                    (backendConfig["qpu-architecture"] + std::string(".txt'")));
+    if (!emulation_mode) {
+      // Use the dynamic quantum architecture of the configured IQM server
+      fetchQuantumArchitecture();
+      pathToFile = writeQuantumArchitectureFile();
+    } else {
+      // In emulation mode there is no server information so no mapping file
```

Comment:
Seeing where you are going with this, I think it would be nice if a quantum architecture file could be manually created and supplied as an argument instead of --iqm-machine. If that argument is provided, then it should take precedence over the server information. That argument should be used for emulation, and a warning should be printed if emulation is used but no architecture file has been specified.